File: ts/test/chat/privacy/mongo/access.ts
import { ObjectId } from "mongodb";
import { MongoLocator } from "../../../../src/model";
import GroupChat from "../../model/gc";
import Message from "../../model/message";
import { FirestoreCollections } from "../../model/shared";
import User from "../../model/user";
import DirectMessage from "../../model/dm";

export default function handleAccessMongo(dataSubjectId: string, locator: MongoLocator, obj: any): Record<string, any> {
    switch (locator.dataType) {
        case 'user':
            const user = obj as User;
            // TODO: should we always convert _id to string after reading from database?
            user.id = obj._id.toString();
            return handleAccessUser(dataSubjectId, locator, user);
        case 'groupChat':
            const groupChat = obj as GroupChat;
            groupChat.id = obj._id.toString();
            return handleAccessGroupChat(dataSubjectId, locator, groupChat);
        case 'directMessage':
            const directMessage = obj as DirectMessage;
            directMessage.id = obj._id.toString();
            return handleAccessDirectMessage(dataSubjectId, locator, directMessage);
        case 'message':
            const message = obj as Message;
            message.id = obj._id.toString();
            return handleAccessMessage(dataSubjectId, locator, message);
        default:
            throw new Error(`Data type ${locator.dataType} does not have handleAccess function implemented`);
    }
}

function handleAccessGroupChat(dataSubjectId: string, locator: MongoLocator, obj: GroupChat): Record<string, any> {
    return {
        messages: {
            dataType: 'message',
            singleDocument: false,
            collection: FirestoreCollections.Messages,
            filter: {
                userID: dataSubjectId,
                chatID: obj.id
            }
        } as MongoLocator
    };
};

function handleAccessUser(dataSubjectId: string, locator: MongoLocator, obj: User): Record<string, any> {
    if (dataSubjectId !== obj.id) {
        return {
            name: obj.name,
        }
    }

    return {
        name: obj.name,
        groupChats: obj.gcs.map((gc: string): MongoLocator => {
            return {
                dataType: 'groupChat',
                singleDocument: true,
                collection: FirestoreCollections.GroupChat,
                filter: {
                    _id: new ObjectId(gc)
                }
            }
        }),
        directMessages: Object.values(obj.dms).map((dm: string): MongoLocator => {
            return {
                dataType: 'directMessage',
                singleDocument: true,
                collection: FirestoreCollections.DirectMessages,
                filter: {
                    _id: new ObjectId(dm)
                }
            }
        })
    };
}

function handleAccessMessage(dataSubjectId: string, locator: MongoLocator, obj: Message): Record<string, any> {
    return {
        content: obj.content,
        timestamp: obj.timestamp
    };
};

function handleAccessDirectMessage(dataSubjectId: string, locator: MongoLocator, obj: DirectMessage): Record<string, any> {
    let otherUserId: string;
    if (obj.user1 === dataSubjectId) {
        otherUserId = obj.user2;
    } else {
        otherUserId = obj.user1;
    }

    return {
        otherUser: {
            dataType: 'user',
            singleDocument: true,
            collection: FirestoreCollections.Users,
            filter: {
                _id: new ObjectId(otherUserId)
            }
        } as MongoLocator,
        messages: {
            dataType: 'message',
            singleDocument: false,
            collection: FirestoreCollections.Messages,
            filter: {
                userID: dataSubjectId,
                chatID: obj.id
            }
        } as MongoLocator
    };
}

File: go/internal/test/chat/access.go
package chat

import (
	"log"

	pal "github.com/privacy-pal/privacy-pal/go/pkg"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

const (
	UserDataType          = "user"
	GroupChatDataType     = "groupchat"
	DirectMessageDataType = "directmessage"
	MessageDataType       = "message"
)

func HandleAccess(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) map[string]interface{} {
	switch currentDbObjLocator.DataType {
	case UserDataType:
		return handleAccessUser(dataSubjectId, currentDbObjLocator, dbObj)
	case GroupChatDataType:
		return handleAccessGroupChat(dataSubjectId, currentDbObjLocator, dbObj)
	case MessageDataType:
		return handleAccessMessage(dataSubjectId, currentDbObjLocator, dbObj)
	case DirectMessageDataType:
		return handleAccessDirectMessage(dataSubjectId, currentDbObjLocator, dbObj)
	default:
		// TODO: should return error
		return nil
	}
}

func handleAccessUser(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) map[string]interface{} {
	data := make(map[string]interface{})

	// TODO: include in documentation: you can access the id in 2 ways
	if dbObj["_id"].(string) != dataSubjectId {
		data["Name"] = dbObj["name"]
		return data
	}
	// if currentDbObjLocator.DocIDs[len(currentDbObjLocator.DocIDs)-1] != dataSubjectId {
	// 	data["Name"] = dbObj["name"]
	// 	return data
	// }

	data["Name"] = dbObj["name"]
	data["Groupchats"] = make([]pal.Locator, 0)
	for _, id := range dbObj["gcs"].([]interface{}) {
		id := id.(string)
		objectID, err := primitive.ObjectIDFromHex(id)
		if err != nil {
			log.Println(err)
		}

		data["Groupchats"] = append(data["Groupchats"].([]pal.Locator), pal.Locator{
			LocatorType: pal.Document,
			DataType:    GroupChatDataType,
			FirestoreLocator: pal.FirestoreLocator{
				CollectionPath: []string{"gcs"},
				DocIDs:         []string{id},
			},
			MongoLocator: pal.MongoLocator{
				Collection: "gcs",
				Filter:     bson.D{{Key: "_id", Value: objectID}},
			},
		})
	}
	data["DirectMessages"] = make([]pal.Locator, 0)
	// TODO: support this in yaml
	for _, dmID := range dbObj["dms"].(map[string]interface{}) {
		dmID := dmID.(string)
		dmIDObj, err := primitive.ObjectIDFromHex(dmID)
		if err != nil {
			log.Println(err)
		}

		data["DirectMessages"] = append(data["DirectMessages"].([]pal.Locator), pal.Locator{
			LocatorType: pal.Document,
			DataType:    DirectMessageDataType,
			FirestoreLocator: pal.FirestoreLocator{
				CollectionPath: []string{"dms"},
				DocIDs:         []string{dmID},
			},
			MongoLocator: pal.MongoLocator{
				Collection: "dms",
				Filter:     bson.D{{Key: "_id", Value: dmIDObj}},
			},
		})
	}

	return data
}

func handleAccessGroupChat(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) map[string]interface{} {
	data := make(map[string]interface{})

	data["Messages"] = pal.Locator{
		LocatorType: pal.Collection,
		DataType:    MessageDataType,
		FirestoreLocator: pal.FirestoreLocator{
			CollectionPath: append(currentDbObjLocator.FirestoreLocator.CollectionPath, "messages"),
			DocIDs:         currentDbObjLocator.DocIDs,
			Filters: []pal.Filter{
				{
					Path:  "userId",
					Op:    "==",
					Value: dataSubjectId,
				},
			},
		},
		MongoLocator: pal.MongoLocator{
			Collection: "messages",
			Filter:     bson.D{{Key: "userId", Value: dataSubjectId}, {Key: "chatId", Value: dbObj["_id"]}},
		},
	}

	return data
}

func handleAccessMessage(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) map[string]interface{} {
	data := make(map[string]interface{})

	data["Content"] = dbObj["content"]
	data["Timestamp"] = dbObj["timestamp"]

	return data
}

func handleAccessDirectMessage(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) map[string]interface{} {
	data := make(map[string]interface{})

	var otherUserId string
	if dbObj["user1"].(string) == dataSubjectId {
		otherUserId = dbObj["user2"].(string)
	} else {
		otherUserId = dbObj["user1"].(string)
	}
	otherUserIDObj, err := primitive.ObjectIDFromHex(otherUserId)
	if err != nil {
		log.Println(err)
	}

	data["Other User"] = pal.Locator{
		LocatorType: pal.Document,
		DataType:    UserDataType,
		FirestoreLocator: pal.FirestoreLocator{
			CollectionPath: []string{"users"},
			DocIDs:         []string{otherUserId},
		},
		MongoLocator: pal.MongoLocator{
			Collection: "users",
			Filter:     bson.D{{Key: "_id", Value: otherUserIDObj}},
		},
	}
	data["Messages"] = pal.Locator{
		LocatorType: pal.Collection,
		DataType:    MessageDataType,
		FirestoreLocator: pal.FirestoreLocator{
			CollectionPath: append(currentDbObjLocator.FirestoreLocator.CollectionPath, "messages"),
			DocIDs:         currentDbObjLocator.DocIDs,
			Filters: []pal.Filter{
				{
					Path:  "userId",
					Op:    "==",
					Value: dataSubjectId,
				},
			},
		},
		MongoLocator: pal.MongoLocator{
			Collection: "messages",
			Filter:     bson.D{{Key: "userId", Value: dataSubjectId}, {Key: "chatId", Value: dbObj["_id"]}},
		},
	}

	return data
}
